### PR TITLE
ci: run Linux release builds in sysroot with Ubuntu 18.04 and LLVM

### PR DESCRIPTION
**Pull request by piscisaureus**
~Depends on: https://github.com/denoland/rusty_v8/pull/708~

### CODE REVIEW

**Comment by piscisaureus** [comment]
| **Before**<sup>1</sup> | **After**<sup>1</sup> |
|:-|:-|
| image | (TBD) |




<sup>1</sup>: Github Actions caches are empty.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -287,28 +385,61 @@ jobs:
...
-          deno run --unstable --allow-write --allow-read --allow-net --allow-env --allow-run ./tools/wpt.ts run --quiet --release --json=wpt.json --wptreport=wptreport.json
-          gzip ./wptreport.json
+          ./target/release/deno run --allow-env --allow-net --allow-read \
+                                    --allow-run --allow-write --unstable \
+                                ./tools/wpt.ts setup
+          ./target/release/deno run --allow-env --allow-net --allow-read     \
+                                    --allow-run --allow-write --unstable     \
+                                ./tools/wpt.ts run --no-build --quiet        \
+                                                   --release --json=wpt.json \
+                                                   --wptreport=wptreport.json
```
Comment (lucacasonato) [inline .github/workflows/ci.yml:442]: So now WPT only run on release, and not on debug? If we keep going with removing WPT targets at this rate, we won't run WPT at all next month...

File: tools/wpt/utils.ts [hunk 2]
```diff
@@ -9,14 +9,16 @@ export const {
...
   release,
+  build,
   rebuild,
   ["--"]: rest,
   ["auto-config"]: autoConfig,
   binary,
 } = parse(Deno.args, {
   "--": true,
-  boolean: ["quiet", "release", "no-interactive"],
+  boolean: ["build", "quiet", "release", "no-interactive"],
```
Comment (lucacasonato) [inline tools/wpt/utils.ts:19]: Use the new `binary` flag for `wpt.ts` instead. If you pass `--binary`, the build will be disabled.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -86,32 +88,20 @@ jobs:
...
-        if: |
-          !startsWith(matrix.os, 'windows')
-        run: |-
+        if: matrix.kind == 'lint'
+        run: |
           curl -fsSL https://deno.land/x/install/install.sh | sh -s v1.7.2
           echo "$HOME/.deno/bin" >> $GITHUB_PATH
 
-      - name: Error on Warning
-        run: echo "RUSTFLAGS=-D warnings" >> $GITHUB_ENV
```
Comment (ry) [inline .github/workflows/ci.yml:96]: Removing this disables error on warning.

This could be cleaner if the following was put above
```
env:
  RUSTFLAGS: -D warnings
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -151,23 +141,92 @@ jobs:
...
-        if: |
-          matrix.kind == 'test' &&
-          matrix.profile == 'release' &&
-          github.repository == 'denoland/deno' &&
-          github.ref == 'refs/heads/main'
-        shell: bash
+      - name: Error on warning
+        # TODO(piscisaureus): enable this on Windows again.
+        if: "!matrix.use_sysroot && !startsWith(matrix.os, 'windows')"
+        run: echo "RUSTFLAGS=-D warnings" >> $GITHUB_ENV
```
Comment (ry) [inline .github/workflows/ci.yml:147]: Oops, I didn't see this.

Why are windows builds getting warnings? I would guess this sysroot change would have no effect on Windows.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -151,23 +141,92 @@ jobs:
...
+          curl https://apt.llvm.org/llvm-snapshot.gpg.key |
+            gpg --dearmor                                 |
+            sudo dd of=/sysroot/etc/apt/trusted.gpg.d/llvm-snapshot.gpg
+          sudo chroot /sysroot apt update -y
+          sudo chroot /sysroot apt install --no-install-recommends -y \
+                                           clang-12 lld-12
+
+          # Make rust available inside the chroot environment.
+          sudo mkdir -p /sysroot/usr/share/rust
+          sudo mount --rbind /usr/share/rust /sysroot/usr/share/rust
```
Comment (ry) [inline .github/workflows/ci.yml:180]: This is so agonizing - all this nonsense to change two symbols in the binary...

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -200,22 +262,45 @@ jobs:
         with:
           path: |
             ./target
-          key: |
-            s0mth1ng_rand0m # Cache never be created for this key.
+            !./target/gn_out
```
Comment (ry) [inline .github/workflows/ci.yml:265]: gn_out is usually at target/debug/gn_out ...

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -200,22 +262,45 @@ jobs:
...
 
+      - name: Shallow clone crates.io index
+        shell: bash
+        run: |
+          if [ ! -d ~/.cargo/registry/index/github.com-1ecc6299db9ec823/.git ]
+          then
+            git clone --depth 1 --no-checkout                      \
+                      https://github.com/rust-lang/crates.io-index \
+                      ~/.cargo/registry/index/github.com-1ecc6299db9ec823
+          fi
```
Comment (ry) [inline .github/workflows/ci.yml:292]: I don't understand this. Is this related to the sysroot?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (piscisaureus) [reply .github/workflows/ci.yml:442]: No? Have you read the entire diff?

Comment (lucacasonato) [reply .github/workflows/ci.yml:442]: Mh, I could have sworn I double checked... I guess not.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -287,28 +381,61 @@ jobs:
...
+        if: |
+          startsWith(matrix.os, 'ubuntu') && matrix.kind == 'test' &&
+          matrix.profile == 'debug'
+        run: |
+          ./target/debug/deno run --allow-env --allow-net --allow-read \
+                                  --allow-run --allow-write --unstable \
+                              ./tools/wpt.ts setup
+          ./target/debug/deno run --allow-env --allow-net --allow-read \
+                                  --allow-run --allow-write --unstable \
+                              ./tools/wpt.ts run --binary --quiet
```
Comment (lucacasonato) [inline .github/workflows/ci.yml:424]: You will need to actually specify a binary (`./target/debug/deno `)

Comment (piscisaureus) [reply .github/workflows/ci.yml:424]: A bit silly to specify `./foo/bar/deno run blah.js --binary ./foo/bar/deno`, no?

Comment (lucacasonato) [reply .github/workflows/ci.yml]: In many scenarios the binary you want to test is not the same one that is coordinating the test.

**Review by piscisaureus — COMMENTED** [review]
_removed_

File: .github/workflows/ci.yml (hunk 7 above)
Comment (piscisaureus) [reply .github/workflows/ci.yml:292]: Separated out the commit

Comment (ry) [reply .github/workflows/ci.yml:292]: What does this do tho? What is "1ecc6299db9ec823" ?

File: .github/workflows/ci.yml (hunk 8 above)
Comment (piscisaureus) [reply .github/workflows/ci.yml]: The --binary option does not work in release mode.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (piscisaureus) [reply .github/workflows/ci.yml:292]: > What does this do tho?

As reflected in the name, it makes a shallow clone of the crates.io index.

> What is "1ecc6299db9ec823"

I don't know.

> Will it ever need to be updated?

I don't know. But in my Rust-using career it has never changed.

> I think this could have a comment as to the purpose

The purpose is to make CI faster. Want me add `# This makes CI faster`?

Comment (ry) [reply .github/workflows/ci.yml:296]:
```
# Manually clone the cargo crate index shallowly so that cargo doesn't
# clone it deeply. This saves about 1 minute of CI time. We don't believe the
# identifier '1ecc6299db9ec823' will ever change, but if it does then this command
# must be updated.
```

**Review by ry — APPROVED** [review]
LGTM
